2.4-representable-functors.md #15

arioston opened the pull request:
This is a draft PR

arioston committed:
2.4-representable-functors.md #15

This is a draft PR

arioston committed:
Merge branch 'master' of https://github.com/arrow-kt/Category-Theory-for-Programmers.kt

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -2,30 +2,34 @@
 ```
 ```kotlin:ank:silent
-
+typealias Reader<A, X> = (A) -> X
````
This declaration is more in line with the `Arrow Code Base` and clearer for the audiance:
```suggestion
typealias Reader<D, A> = Kleisli<ForId, D, A>
```
We can Swap the variables in the book later

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -2,30 +2,34 @@
 ```
 ```kotlin:ank:silent
-
+typealias OP<A,X> = (X)-> A
````
```suggestion
typealias Op<A, X> = (X) -> A
```
check your indent's.

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -2,30 +2,34 @@
+```kotlin:ank:playground
+import arrow.core.compose
+ 
+data class Op<A, B> constructor(private val f: (A) -> B) {
````
This is not how you write typeclass-instances, try to do it as in the example code you wrote below:
Here is an example for Kleisli
```
interface KleisliContravariant<F, D> : Contravariant<Conested<Kind<ForKleisli, F>, D>> {
  override fun <A, B> Kind<Conested<Kind<ForKleisli, F>, D>, A>.contramap(f: (B) -> A): Kind<Conested<Kind<ForKleisli, F>, D>, B> =
    counnest().fix().local(f).conest()

  fun <A, B> KleisliOf<F, A, D>.contramapC(f: (B) -> A): KleisliOf<F, B, D> =
    conest().contramap(f).counnest()
}
```

arioston replied:
I saw in the section [natural transformations](https://github.com/arrow-kt/Category-Theory-for-Programmers.kt/blob/master/src/main/ank/1.10-natural-transformations.md) :
```haskell
instance Contravariant (Op r) where
  contramap f (Op g) = Op (g . f)
```
```kotlin
import arrow.core.compose

data class Op<A, B> constructor(private val f: (A) -> B) {
  fun <C> contramap(op: Op<B, C>) = Op<A, C> { (op.f compose f)(it) }
}
```
are they different?

i-walker replied:
Yes drastically. 
Refer to your implementation a few lines below. 
PIng me in Slack if you need help

i-walker replied:
@arioston  Thank you for pointing out the error [here](https://github.com/arrow-kt/Category-Theory-for-Programmers.kt/blob/master/src/main/ank/1.10-natural-transformations.md). I will correct it in the next PR

arioston replied:
@i-walker is this the correct implementation ?
```kotlin
interface OpContravariant<C>: Contravariant<Conested<ForOp, C>> {
    override fun <A, B> Kind<Conested<ForOp, C>, A>.contramap(f: (B) -> A): Kind<Conested<ForOp, C>, B> =
         counnest().fix().compose(f.k()).conest()
}
```

i-walker replied:
This may look like this:
```
@higherkind
data class Op<A, X>(val f: (X) -> A): OpOf<A, X> {
  fun <B> compose(other: Op<B, A>): Op<B, X> =
    Op(other.f compose f)
  fun <B> compose(g: (A) -> B): Op<B, X> =
    compose(Op(g))
}
```
I wrote 2 isomorphic representations of `compose`. You can choose, which one makes more sense. I prefer the one without the wrapper

i-walker replied:
Thanks to @1Jajen1 the contravariant instance may look like this:
```
interface OpContravariant<I>: Contravariant<OpPartialOf<I>> {
  override fun <A, B> Kind<OpPartialOf<I>, A>.contramap(f: (B) -> A): Kind<OpPartialOf<I>, B> =
    Op(fix().f compose f)
}
```

i-walker commented on drafts/2.4-representable-functors.md:
```diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
-
+import arrow.extension
+import arrow.typeclasses.Profunctor
+@extension
```
Add space
```suggestion

@extension
```

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
     rmap = (.)
 ```
 ```kotlin:ank:silent
````
This should preferably be in a playground env.
```suggestion
```kotlin:ank:playground
```

i-walker commented on drafts/2.4-representable-functors.md:
```diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
+import arrow.extension
+import arrow.typeclasses.Profunctor
+@extension
+interface ArrowProfunctor: Profunctor<ForFunction1> {
```
```suggestion
interface Function1Profunctor: Profunctor<ForFunction1> {
```

i-walker commented on drafts/2.4-representable-functors.md:
```diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
+        (fr compose fix().f compose fl).k()
+ 
+    override fun <A, B, C> Kind2<ForFunction1, A, B>.lmap(f: (C) -> A): Kind2<ForFunction1, C, B> =
+        (f andThen fix().f).k()
```
Express `lmap` and `rmap` in terms of dimap, so readers understand better that dimap is sufficient enough to get a Profunctor instance. You can see the impl in `Arrow`

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
 ```
-```kotlin:ank:silent
-
+```kotlin:kotlin
````
`kotlin:kotlin` doesn't exist. If you don't want to run the snippet you solely put `kotlin`
```suggestion
```kotlin
```

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
-```kotlin:ank:silent
-
+```kotlin:kotlin
+fun<A, F, G> alpha(): (Kind<F, A>) -> Kind<G, A>
````
indents
```suggestion
fun <A, F, G> alpha(): (Kind<F, A>) -> Kind<G, A>
```

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
 ```
 ```kotlin:ank:silent
-
+fun<A, F, G> beta(): (Kind<F, A>) -> Kind<G, A>
````
```suggestion
fun <A, F, B> beta(): (Kind<F, A>) -> ((B) -> A)
```

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
 ```
-```kotlin:ank:silent
-
+```kotlin
````
turn this into a playground env.

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -35,50 +39,62 @@ instance Profunctor (->) where
-```kotlin:ank:silent
-
+```kotlin
+fmap(f)(fmap(h)(listOf(12))) == fmap(f compose h)(listOf(12))
````
These would be `map` functions not `fmap`, because we assume the functions `h` and `f`  go from some `(Y) -> Z` and we stay in the List container.
Both ways work, but map fits more to the haskell code

arioston replied:
should I change the fmap name to map?
```Haskell
instance Functor (Reader a) where
    fmap f h = f . h
```
```kotlin:ank:silent
fun <D, A> fmap(run: (D) -> A): Reader<D, A> = Kleisli(run.andThen { Id(it) })
```

i-walker replied:
Come again?

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -95,14 +111,17 @@ class Representable f where
     index    :: f x -> Rep f -> x
 ```
 ```kotlin:ank:silent
````
typeclasses need to be in a playground
```suggestion
```kotlin:ank:playground
```

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -95,14 +111,17 @@ class Representable f where
 ```
 ```kotlin:ank:silent
-
+interface Representable<F, A> {
````
fix the indents in your body and the type signutares are wrong. Representable is a Typeconstructor. It solely needs the `F`
```suggestion
interface Representable<F> {
```

i-walker reviewed (commented):
Nice having you back @arioston :)

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
Update drafts/2.4-representable-functors.md

Co-Authored-By: Imran Settuba <46971368+i-walker@users.noreply.github.com>

arioston committed:
2.4-representable-functors.md #15

1. Implementation of lmap and rmap was changed to use dimap
2. fmap on list was changed to use map instead..

arioston committed:
Merge remote-tracking branch 'origin/master'

i-walker commented on drafts/2.4-representable-functors.md:
````diff
@@ -2,30 +2,34 @@
 ```
 ```kotlin:ank:silent
-
+typealias Op<A, X> = (X) -> A
````
I suggest using Op<A, X> as a data class, because we're using it in a Higherkinded position in `OpContravariant`. But your right to define it as a type alias.
Additionally, you may define the hkt version in the dedicated folder src/kotlin/higherkinds with the annotation, so your playground can resolve the import for `ForOp`.

i-walker reviewed (approved):
Looks good to me :) 
🎉 
Just a view changes here and there.

arioston committed:
2.4-representable-functors.md #15

1. Creation of data class Op in higherkinds package.
2. Change contravariant implementation to use @1Jajen1's implementation.
3. Implementation of beta.

arioston committed:
Move From Draft to Ank

1. Move 2.4-representable-functors.md from draft to Ank

arioston committed:
2.4-representable-functors.md #15

1. Creation of data class Stream in higherkinds package.
2. Addition of imports
3. Change Representable interface to receive two type parameters F, A.

i-walker commented on src/main/ank/2.4-representable-functors.md:
````diff
@@ -0,0 +1,163 @@
+```kotlin:ank:playground
+import arrow.syntax.function.compose
+
+val streamRepresentable = object: Representable<ForStream, Int>{
````
The Stream data type in the book references an effectfull data type Stream, which does not exist in ArrowFx, yet.

i-walker reviewed (commented):
Looks good to me. Please, consider the note I left about the Stream instance and remove the data type.

arioston committed:
2.4-representable-functors.md #15

1. Removing the Stream Data Type

i-walker commented:
Thanks @arioston 🎉
